File: src/main/ruby/template_builder.rb
####################################################################################################
#
# Copyright 2013-2019 ISP RAS (http://www.ispras.ru)
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
# in compliance with the License. You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License
# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
# or implied. See the License for the specific language governing permissions and limitations under
# the License.
#
####################################################################################################

require_relative 'template'

# TODO: make errors display file/line numbers (there's an example of that in the last version).

module TemplateBuilder

  def self.define_runtime_methods(metamodel)
    modes = metamodel.getAddressingModes
    modes.each { |mode| define_addressing_mode mode }

    mode_groups = metamodel.getAddressingModeGroups
    mode_groups.each { |mode_group| define_addressing_mode_group mode_group.getName().to_s }

    ops = metamodel.getOperations
    ops.each { |op| define_operation op }

    op_groups = metamodel.getOperationGroups
    op_groups.each { |op_group| define_operation_group op_group.getName().to_s }

    registers = metamodel.getRegisters
    registers.each { |register| define_store register }

    memories = metamodel.getMemoryStores
    memories.each { |memory| define_store memory }
  end

  # Defines methods to access registers and memory (added to the Template class)
  def define_store(store)
    name = store.getName().to_s

    p = lambda do |index|
      location(name, index)
    end

    define_method_for Template, name, "store", p
  end

  # Defines methods for addressing modes (added to the Template class)
  def define_addressing_mode(mode)
    name = mode.getName().to_s
    #puts "Defining mode #{name}..."

    p = lambda do |*arguments, &situations|
      builder = @template.newAddressingModeBuilder name
      set_arguments builder, arguments

      if situations != nil
        builder.setSituation @situation_manager.instance_eval &situations
      else
        default_situation = @template.getDefaultSituation name
        if default_situation != nil
          builder.setSituation default_situation
        end
      end

      builder.build
    end

    define_method_for Template, name, "mode", p
  end

  # Defines methods for addressing mode groups (added to the Template class)
  def define_addressing_mode_group(name)
    #puts "Defining mode group #{name}..."

    p = lambda do |*arguments, &situations|
      builder = @template.newAddressingModeBuilderForGroup name
      set_arguments builder, arguments

      if situations != nil
        builder.setSituation @situation_manager.instance_eval &situations
      else
        default_situation = @template.getDefaultSituation group_name.to_s

        if default_situation == nil
          default_situation = @template.getDefaultSituation name
        end

        if default_situation != nil
          builder.setSituation default_situation
        end
      end

      builder.build
    end

    define_method_for Template, name, "mode", p
  end

  # Defines methods for operations (added to the Template class)
  def define_operation(op)
    name = op.getName().to_s
    #puts "Defining operation #{name}..."

    is_root = op.isRoot
    root_shortcuts = op.hasRootShortcuts

    p = lambda do |*arguments, &situations|
      builder = @template.newOperationBuilder name
      set_arguments builder, arguments

      if situations != nil
        builder.setSituation @situation_manager.instance_eval &situations
      else
        default_situation = @template.getDefaultSituation name
        if default_situation != nil
          builder.setSituation default_situation
        end
      end

      if is_root
        @template.setRootOperation builder.build, get_caller_location
        @template.endBuildingCall
      elsif root_shortcuts
        # TODO: Dirty hack! Assumes that if a root shortcut exists, we always use it.
        builder.setContext "#root"
        @template.setRootOperation builder.build, get_caller_location
        @template.endBuildingCall
      else
        builder
      end
    end

    define_method_for Template, name, "op", p
  end

  # Defines methods for operation groups (added to the Template class)
  def define_operation_group(group_name)
    #puts "Defining operation group #{group_name}..."

    p = lambda do |*arguments, &situations|
      op = @template.chooseMetaOperationFromGroup group_name
      name = op.getName().to_s

      is_root = op.isRoot
      root_shortcuts = op.hasRootShortcuts

      builder = @template.newOperationBuilder name
      set_arguments builder, arguments

      if situations != nil
        builder.setSituation @situation_manager.instance_eval &situations
      else
        default_situation = @template.getDefaultSituation group_name.to_s

        if default_situation == nil
          default_situation = @template.getDefaultSituation name
        end

        if default_situation != nil
          builder.setSituation default_situation
        end
      end

      if is_root
        @template.setRootOperation builder.build, get_caller_location
        @template.endBuildingCall
      elsif root_shortcuts
        # TODO: Dirty hack! Assumes that if a root shortcut exists, we always use it.
        builder.setContext "#root"
        @template.setRootOperation builder.build, get_caller_location
        @template.endBuildingCall
      else
        builder
      end
    end

    define_method_for Template, group_name, "op", p
  end

  def set_arguments(builder, args)
    if args.count == 1 and args.first.is_a?(Hash)
      set_arguments_from_hash builder, args.first
    else
      set_arguments_from_array builder, args
    end
  end

  def set_arguments_from_hash(builder, args)
    args.each_pair do |name, value|
      value = value.java_object if value.is_a? WrappedObject
      value = value.to_s if value.is_a? Symbol
      builder.setArgument name.to_s, value
    end
  end

  def set_arguments_from_array(builder, args)
    args.each do |value|
      if value.is_a? Array
        set_arguments_from_array builder, value
      else
        value = value.java_object if value.is_a? WrappedObject
        value = value.to_s if value.is_a? Symbol
        builder.addArgument value
      end
    end
  end

  # Defines a method in the target class.
  #
  # If such method is already defined, the method type is  added to the method name as a prefix
  # to make the name unique. If this does not help, an error is reported.
  #
  # Parameters:
  #   target_class Target class (Class object)
  #   method_name  Method name (String)
  #   method_type  Method type (String)
  #   method_body  Body for the method (Proc)
  def define_method_for(target_class, method_name, method_type, method_body)
    method_name = method_name.downcase
    # puts "Defining method #{target_class}.#{method_name} (#{method_type})..."

    if !target_class.method_defined?(method_name)
      target_class.send(:define_method, method_name, method_body)
    elsif !target_class.method_defined?("#{method_type}_#{method_name}")
      target_class.send(:define_method, "#{method_type}_#{method_name}", method_body)
    else
      puts "Error: Failed to define the #{method_name} method (#{method_type})"
    end
  end

end # TemplateBuilder

File: src/main/python/template_builder.py
#
# Copyright 2018 ISP RAS (http://www.ispras.ru)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

#from ru.ispras.microtesk.test import TestEngine
#engine = TestEngine.getInstance()
#define_runtime_methods(engine.getModel().getMetaData())


import template
import globals
#import importlib
#import ntpath

#dummy = ntpath.basename(globals.TEMPLATE_FILE[:-3])
#template_e = importlib.import_module(dummy)

#print template_e.template


from java.math import BigInteger



def define_runtime_methods(metamodel):
    
    modes = metamodel.getAddressingModes()
    for x in modes:
        define_addressing_mode(x)
    
    mode_groups = metamodel.getAddressingModeGroups()
    for x in mode_groups:
        define_addressing_mode_group(x.getName())
        
    ops = metamodel.getOperations()
    for x in ops:
        define_operation(x)
        
    op_groups = metamodel.getOperationGroups()
    for x in op_groups:
        define_operation_group(x.getName())
        
    registers = metamodel.getRegisters()
    for x in registers:
        define_store(x)
    
    memories = metamodel.getMemoryStores()
    for x in memories:
        define_store(x)
 

def define_store(store):
  name = store.getName()

  def p(index):
    return template.location(name, index)

  define_method_for(template, name, "store", p)
        
def define_addressing_mode(mode):
    
    name  = mode.getName()
    
    #print "Defining mode {}...".format(name)
    
    
    def p(*arguments, **kwargs):
        
        
        builder = globals.template.template.newAddressingModeBuilder(name)
        set_argumets(builder, arguments)
        situations = kwargs.get('situations')
        
        if situations != None:
            pass
        else:
            default_situation = globals.template.template.getDefaultSituation(name)
            if default_situation != None:
                builder.setSituation(default_situation)
            
        return builder.build()
    
    define_method_for(template, name, "mode", p)
    
def define_addressing_mode_group(name):
    
    #print "Defining mode group {}...".format(name)
    
    def p(*arguments, **kwarg):
        builder = globals.template.template.newAddressingModeBuilderForGroup(name)
        set_argumets(builder, arguments)
        situations = kwargs.get('situations')
        
        if situations != None:
            pass
        else:
            default_situation = globals.template.template.getDefaultSituation(group_name)
            if default_situation == None:
                default_situation = globals.template.template.getDefaultSituation(name)
             
            if default_situation != None:
                builder.setSituation(default_situation) 
        
        return builder.build()
    
    define_method_for(template, name, "mode", p)

def define_operation(op):
    name = op.getName()
    
    #print "Defining operation {}...".format(name)
    
    is_root = op.isRoot()
    root_shortcuts = op.hasRootShortcuts()
    def p(*arguments, **kwargs):
        builder = globals.template.template.newOperationBuilder(name)
        set_argumets(builder, arguments)
        situations = kwargs.get('situations')
        
        
        if situations != None:
            pass
        else:
            default_situation = globals.template.template.getDefaultSituation(name)
            if default_situation != None:
                builder.setSituation(default_situation)
                
        if is_root:
            globals.template.template.setRootOperation(builder.build(),globals.template.get_caller_location())
            globals.template.template.endBuildingCall()
        elif root_shortcuts:
            builder.setContext("#root")
            globals.template.template.setRootOperation(builder.build(),globals.template.get_caller_location())
            globals.template.template.endBuildingCall()
            
        else:
            return builder
        
    define_method_for(template, name, "op", p)
    
def define_operation_group(group_name):
    
    #print "Defining operation group {}...".format(group_name)
    
    def p(*arguments, **kwargs):
        op = globals.template.template.chooseMetaOperationFromGroup(group_name)
        name = op.getName()
        situations = kwargs.get('situations')
    
        is_root = op.isRoot()
        root_shortcuts = op.hasRootShortcuts()
        
        builder = globals.template.template.newOperationBuilder(name)
        set_argumets(builder, arguments)
        
        
        if situations != None:
            pass
        else:
            default_situation = globals.template.template.getDefaultSituation(group_name)
            if default_situation == None:
                default_situation = globals.template.template.getDefaultSituation(name)
            
            if default_situation != None:
                builder.setSituation(default_situation)
                
        if is_root:
            globals.template.template.setRootOperation(builder.build(),globals.template.get_caller_location())
            globals.template.template.endBuildingCall()
        elif root_shortcuts:
            builder.setContext("#root")
            globals.template.template.setRootOperation(builder.build(),globals.template.get_caller_location())
            globals.template.template.endBuildingCall()
            
        else:
            return builder
        
    define_method_for(template, group_name, "op", p)
        
    
def set_argumets(builder,args): 
    if len(args) == 1 and type(args[0]) is dict:
        set_arguments_from_hash(builder, args.first())
    else:
        set_arguments_from_array(builder, args)
        
        
def set_arguments_from_hash(builder,args):
    for name, value in args.iteritems():
        if isinstance(value, template.WrappedObject):
            value = value.java_object 
        if isinstance(value, basestring):
            value = value
        if isintance(value,int):
            builder.setArgument(name,BigInteger(str(value)))
        else:
            builder.setArgument(name,value)
        
        
def set_arguments_from_array(builder,args):
    for value in args:
        if type(args) is list:
            set_arguments_from_array(builder, value)
        else:
            if isinstance(value, template.WrappedObject):
                value = value.java_object 
            if isinstance(value, basestring):
                value = value
            if isinstance(value,int):
                builder.addArgument(BigInteger(str(value)))
            else:
                builder.addArgument(value)
    
        
def define_method_for(target_class, method_name, method_type, method_body):
    method_name = method_name.lower()
    
    if method_name == "or" or method_name == "and":
        method_name = method_name.upper()
    
    print "Defining method {}.{}".format(target_class,method_name)
    
    if not hasattr(target_class, method_name):
        #setattr(target_class,method_name,MethodType(method_body,None,target_class))
        setattr(target_class, method_name, method_body)
    elif not hasattr(target_class, "{}_{}".format(method_type,method_name)):
     #   setattr(target_class,"{}_{}".format(method_type,method_name),MethodType(method_body,None,target_class))
        setattr(target_class,"{}_{}".format(method_type,method_name),method_body)
    else:
        print "Error: Failed to define the {} method ({})".format(method_type,method_name)
    
    
    
    
    
    
    
    
    
    
    
    
    